File: euhatrtsplib/src/main/java/com/euhat/rtsp/euhatrtsplib/EuhatRtspPlayer.java
/*
 *  EuhatRtsp
 *  library and sample to play and acquire rtsp stream on Android device
 *
 * Copyright (c) 2014-2018 Euhat.com euhat@hotmail.com
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 *  All files in the folder are under this Apache License, Version 2.0.
 *  Files in the libjpeg-turbo, libusb, libuvc, rapidjson folder
 *  may have a different license, see the respective files.
 */

package com.euhat.rtsp.euhatrtsplib;

import java.io.File;
import java.io.RandomAccessFile;
import java.util.ArrayList;
import java.util.List;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.graphics.SurfaceTexture;
import android.text.TextUtils;
import android.util.Log;
import android.view.Surface;
import android.view.SurfaceHolder;

public class EuhatRtspPlayer {
	private static final boolean DEBUG = false;
	private static final String TAG = EuhatRtspPlayer.class.getSimpleName();

	public static final int DEFAULT_PREVIEW_WIDTH = 1920;
	public static final int DEFAULT_PREVIEW_HEIGHT = 1080;
	public static final int DEFAULT_PREVIEW_FPS = 25;

	public static final int PIXEL_FORMAT_YV12 = 1;
	public static final int PIXEL_FORMAT_NV21 = 2;

	static {
		System.loadLibrary("EuhatRtsp");
	}

	protected int mCurrentWidth = DEFAULT_PREVIEW_WIDTH, mCurrentHeight = DEFAULT_PREVIEW_HEIGHT;
    protected long mNativePtr;

    public EuhatRtspPlayer() {
    	mNativePtr = nativeCreate();
	}

    public synchronized void open(String url, int rtspTimeOut, int rtspFlag, int reconnectTimeOut, int displayFps, int displayBufCount) {
    	int result;
    	try {
			result = nativeConnect(mNativePtr, url, rtspTimeOut, rtspFlag, reconnectTimeOut, displayFps, displayBufCount);
		} catch (final Exception e) {
			Log.w(TAG, e);
			result = -1;
		}
		if (result != 0) {
			throw new UnsupportedOperationException("open failed, result is " + result);
		}
    }

    public void open(String url) {
    	open(url, 8000, 1, 10000, 25, 100);
	}

    public synchronized void close() {
    	stopPreview();
    	if (mNativePtr != 0) {
    		nativeClose(mNativePtr);
    	}
    	if (DEBUG) Log.v(TAG, "closed.");
    }

    public synchronized void setPreviewDisplay(final SurfaceHolder holder) {
   		nativeSetPreviewDisplay(mNativePtr, holder.getSurface());
    }

    public synchronized void setPreviewTexture(final SurfaceTexture texture) {
    	final Surface surface = new Surface(texture);	// XXX API >= 14
    	nativeSetPreviewDisplay(mNativePtr, surface);
    }

    public synchronized void setPreviewDisplay(final Surface surface) {
    	nativeSetPreviewDisplay(mNativePtr, surface);
    }

    public void setStatusCallback(final IStatusCallback callback) {
    	nativeSetStatusCallback(mNativePtr, callback);
	}

    public void setFrameCallback(final IFrameCallback callback, final int pixelFormat) {
       	nativeSetFrameCallback(mNativePtr, callback, pixelFormat);
    }

    public synchronized void startPreview() {
   		nativeStartPreview(mNativePtr);
    }

    public synchronized void stopPreview() {
    	setFrameCallback(null, 0);
    	nativeStopPreview(mNativePtr);
    }

    public synchronized void destroy() {
    	close();
    	if (mNativePtr != 0) {
    		nativeDestroy(mNativePtr);
    		mNativePtr = 0;
    	}
    }

    private final native long nativeCreate();
    private final native void nativeDestroy(long euhatRtsp);

    private final native int nativeConnect(long euhatRtsp, String url, int rtspTimeOut, int rtspFlag, int reconnectTimeOut, int displayFps, int displayBufCount);
    private static final native int nativeClose(long euhatRtsp);

    private static final native int nativeStartPreview(long euhatRtsp);
    private static final native int nativeStopPreview(long euhatRtsp);
    private static final native int nativeSetPreviewDisplay(long euhatRtsp, final Surface surface);
    private static final native int nativeSetStatusCallback(long euhatRtsp, final IStatusCallback callback);
    private static final native int nativeSetFrameCallback(long euhatRtsp, final IFrameCallback callback, int pixelFormat);

}

File: euhatrtsplib/src/main/java/com/euhat/rtsp/euhatrtsplib/EuhatRtspPlayer.kt
/*
 *  EuhatRtsp
 *  library and sample to play and acquire rtsp stream on Android device
 *
 * Copyright (c) 2014-2018 Euhat.com euhat@hotmail.com
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 *  All files in the folder are under this Apache License, Version 2.0.
 *  Files in the libjpeg-turbo, libusb, libuvc, rapidjson folder
 *  may have a different license, see the respective files.
 */
package com.euhat.rtsp.euhatrtsplib

import android.graphics.SurfaceTexture
import android.util.Log
import android.view.Surface
import android.view.SurfaceHolder

class EuhatRtspPlayer {
    companion object {
        private const val DEBUG = false
        private val TAG = EuhatRtspPlayer::class.java.simpleName
        const val DEFAULT_PREVIEW_WIDTH = 1920
        const val DEFAULT_PREVIEW_HEIGHT = 1080
        const val DEFAULT_PREVIEW_FPS = 25
        const val PIXEL_FORMAT_YV12 = 1
        const val PIXEL_FORMAT_NV21 = 2

        init {
            try {
                System.loadLibrary("EuhatRtsp")
            } catch (ex:Throwable) {
                ex.printStackTrace()
            }
        }
    }

    protected var mCurrentWidth = DEFAULT_PREVIEW_WIDTH
    protected var mCurrentHeight = DEFAULT_PREVIEW_HEIGHT
    protected var mNativePtr: Long
    @Synchronized
    fun open(url: String, rtspTimeOut: Int, rtspFlag: Int, reconnectTimeOut: Int, displayFps: Int, displayBufCount: Int) {
        val result: Int
        result = try {
            nativeConnect(mNativePtr, url, rtspTimeOut, rtspFlag, reconnectTimeOut, displayFps, displayBufCount)
        } catch (e: Exception) {
            Log.w(TAG, e)
            -1
        }
        if (result != 0) {
            throw UnsupportedOperationException("open failed, result is $result")
        }
    }

    fun open(url: String) {
        open(url, 8000, 1, 10000, 25, 100)
    }

    @Synchronized
    fun close() {
        stopPreview()
        if (mNativePtr != 0L) {
            nativeClose(mNativePtr)
        }
        if (DEBUG) Log.v(TAG, "closed.")
    }

    @Synchronized
    fun setPreviewDisplay(holder: SurfaceHolder) {
        nativeSetPreviewDisplay(mNativePtr, holder.surface)
    }

    @Synchronized
    fun setPreviewTexture(texture: SurfaceTexture?) {
        val surface = Surface(texture) // XXX API >= 14
        nativeSetPreviewDisplay(mNativePtr, surface)
    }

    @Synchronized
    fun setPreviewDisplay(surface: Surface) {
        nativeSetPreviewDisplay(mNativePtr, surface)
    }

    fun setStatusCallback(callback: IStatusCallback) {
        nativeSetStatusCallback(mNativePtr, callback)
    }

    fun setFrameCallback(callback: IFrameCallback?, pixelFormat: Int) {
        nativeSetFrameCallback(mNativePtr, callback, pixelFormat)
    }

    @Synchronized
    fun startPreview() {
        nativeStartPreview(mNativePtr)
    }

    @Synchronized
    fun stopPreview() {
        setFrameCallback(null, 0)
        nativeStopPreview(mNativePtr)
    }

    @Synchronized
    fun destroy() {
        close()
        if (mNativePtr != 0L) {
            nativeDestroy(mNativePtr)
            mNativePtr = 0
        }
    }

    private external fun nativeCreate(): Long
    private external fun nativeDestroy(euhatRtsp: Long)
    private external fun nativeConnect(euhatRtsp: Long, url: String, rtspTimeOut: Int, rtspFlag: Int, reconnectTimeOut: Int, displayFps: Int, displayBufCount: Int): Int
    private external fun nativeClose(euhatRtsp: Long): Int
    private external fun nativeStartPreview(euhatRtsp: Long): Int
    private external fun nativeStopPreview(euhatRtsp: Long): Int
    private external fun nativeSetPreviewDisplay(euhatRtsp: Long, surface: Surface): Int
    private external fun nativeSetStatusCallback(euhatRtsp: Long, callback: IStatusCallback): Int
    private external fun nativeSetFrameCallback(euhatRtsp: Long, callback: IFrameCallback?, pixelFormat: Int): Int

    init {
        mNativePtr = nativeCreate()
    }
}